gadgets/trace_dns/program.bpf.c

```diff
@@ -262,21 +262,51 @@ int ig_trace_dns(struct __sk_buff *skb)
 		return 0;
 	}
 
+	// Since we have the same offset for source and destination ports for both TCP and UDP,
+	// we can use the same code to extract them.
+	// - offsetof(struct udphdr, source) == offsetof(struct tcphdr, source)
+	// - offsetof(struct udphdr, dest) == offsetof(struct tcphdr, dest)
 	switch (proto) {
 	case IPPROTO_UDP:
+	case IPPROTO_TCP:
 		sport = load_half(skb,
 				  l4_off + offsetof(struct udphdr, source));
 		dport = load_half(skb, l4_off + offsetof(struct udphdr, dest));
-		dns_off = l4_off + sizeof(struct udphdr);
 		break;
-	// TODO: support TCP
 	default:
 		return 0;
 	}
 
 	if (!is_dns_port(sport) && !is_dns_port(dport))
 		return 0;
 
+	// Calculate the DNS offset in the packet
+	switch (proto) {
+	case IPPROTO_UDP:
+		dns_off = l4_off + sizeof(struct udphdr);
+		break;
+	case IPPROTO_TCP:
+		// This is best effort, since we don't reassemble TCP segments.
+		struct tcphdr tcph;
+		if (bpf_skb_load_bytes(skb, l4_off, &tcph, sizeof tcph))
+			return 0;
+
+		// The data offset field in the header is specified in 32-bit words. We
+		// have to multiply this value by 4 to get the TCP header length in bytes.
+		__u8 tcp_header_len = tcph.doff * 4;
+
+		// Skip if we don't have any data to avoid handling control segments
+		dns_off = l4_off + tcp_header_len;
+		if (skb->len <= dns_off)
+			return 0;
+
+		// DNS is after the TCP header and the 2 bytes of the length of the DNS packet
+		dns_off += 2;
```

Comment:
I'm wondering if we can discard here the fragmented DNS payload by comparing that length field with the data length (`skb->len`). I'd expect that for the fragmented payloads, the length present in this 2-bytes field will be larger than the actual payload in the packet we are parsing: `dns_off + 2 + dns_tcp_lenght` > `skb->len`. WDYT?

My goal is to avoid sending it to user space if it will be anyway discarded by the wasm parsing.

Reply:
Good point. I did some testing and second fragment doesn't contain any length so the above check will only work for first packet. Having said that, having a check to skip if there is mismatch between calculated payload length and actual payload length can help:
```
		// If the calculated length of the DNS packet is not the same as the length of the TCP payload, skip
		// the packet. This is a best effort check, as we don't reassemble TCP segments.
		__u16 dns_payload_len = load_half(skb, dns_off);
		if (dns_payload_len != skb->len - dns_off - 2)
			return 0;
```

Do you think it is worth it?

Reply:
Oh, I see. With this check, we reduce the amount of "_WARN[0034] failed to unpack dns message: Name: segment prefix is reserved_", but didn't remove them completely. To do that, we will need to keep track of the 5-tuple, and I don't know if it is worth it.

@alban any comment?

> Because it might be better if we inform user if certain things aren't being traced
>
It's documented that we don't support it, so, imo, it's fine to silently skip them. Now that I think ... It could be a good counter to export as metric: the number of TCP segments belonging to a DNS query/answer that were discarded by the gadget because they were fragmented over the network. It's more an "internal" information but indicates to users what they are not seeing on the trace.

cc @mauriciovasquezbernal